### ISSUE TITLE
critical css rendering bug with new <link rel="preload"> standard

### ISSUE THREAD

**Issue by Enalmada** [turbolinks/turbolinks#317]
The new <link rel="preload"> standard enables us to load stylesheets asynchronously, without blocking rendering.  (Can read more about it here: https://github.com/filamentgroup/loadCSS)

Unfortunately when using this optimization with turbolinks (on browser that supports it like Chrome), the preloaded css is getting applied in the wrong order.

Summary:
```
<!-- Non blocking css load of non critical stuff (fonts, etc)-->
<link rel="preload" href="HugeCssWithRedBody.css" as="style" onload="this.rel='stylesheet'">
<!-- Inlined critical css that likely overrides 3rd party preloaded defaults-->
<style>body {background: white; }
...
<!-- Turbolinks applies preload css over inlined css -->
<a href="turbolinks.html">TurboLink</a>
```

Here is a example:
   https://www.gelltest.com/turbolinks

**Comment by Enalmada** [comment]
Is this a legitimate turbolinks bug or is there something I can/should be doing different?
I doubt I am the first person to want to use turbolinks and preloading css nonblocking so I apologize in advance if I am being silly about something.

**Comment by brian-kephart** [comment]
It's unclear to me from your description whether the problem is associated with a turbolinks event or simply a result of the preloaded stylesheet being applied after the download finishes.

1. Does the above code work when you don't use turbolinks?
2. Does the behavior change based on whether your browser cache is in use versus disabled? 
3. Does the behavior change based on whether you do a full page load versus a turbolinks navigation?

**Comment by Enalmada** [comment]
@brian-kephart  Thanks for your consideration on this issue.
1) Yes, the code works without turbolinks.  I believe that loading stylesheets async like that is a best practice.  
2) Behavior doesn't seem to change with browser cache enabled disabled
3) Full page load and normal links work fine.  Only turbolinks fail.

Please be sure you see the example here of a bare bones page with a normal link and a turbolink: https://www.gelltest.com/turbolinks

**Comment by brian-kephart** [comment]
I think I have an answer of sorts. Turbolinks preserves your stylesheet as expected, but since the element attributes change after download, Turbolinks doesn't recognize it as the same element when the link is clicked, so it adds the "missing" element of the new page into the `<head>`. The styles aren't just in the wrong order, the stylesheet is present twice in the `<head>` after clicking the link.

First thought: mark the element `data-turbolinks-permanent` as described [here](https://github.com/turbolinks/turbolinks#persisting-elements-across-page-loads). Make sure to give the element an `id` attribute first.

If that doesn't work, try marking the element `data-turbolinks-eval="false"` as described [here](https://github.com/turbolinks/turbolinks#working-with-script-elements). This is intended for `<script>` elements, so I have no idea if it would work for stylesheets, but it's easy to try if the previous solution fails.

If neither solution works, I think this has to be treated as an unsupported action or a bug to be fixed.

**Comment by glebm** [comment]
Neither of these suggestions work.

The `data-turbolinks-eval` one doesn't work because that attribute is not considered in `head_details.coffee`:

https://github.com/turbolinks/turbolinks/blob/1066805f5e827804b4b68288e20664d7ec4b45da/src/turbolinks/head_details.coffee#L4

The `data-turbolinks-permanent` one doesn't work because permanent elements only work in `body`:

https://github.com/turbolinks/turbolinks/blob/1066805f5e827804b4b68288e20664d7ec4b45da/src/turbolinks/snapshot_renderer.coffee#L78-L82

**Comment by brian-kephart** [comment]
Thanks @glebm for those links. Looking at them, It looks to me like it would take some very creative work to make this behavior work out of the box. Plus, tracking tags that mutate after the initial page load seems like a big door to open.

@Enalmada what if you just modified the `onload` attribute? Something like:
`this.insertAdjacentHTML('afterend','<link rel="stylesheet" href="HugeCssWithRedBody.css">')`

**Comment by glebm** [comment]
Yes, you can use `preload` headers instead of tag attributes and then just `loadCSS('#{stylesheet_url(:application)}');` as a workaround.

Note that this used to work with turbolinks-classic as turbolinks-classic didn't check the other attributes.
I believe making it work could as simple as having Turbolinks support something the list of attributes to track (with the default being all), or the list of attributes to ignore when tracking. E.g. `data-turbolinks-track-attributes-only` / `data-turbolinks-track-attributes-exclude`.

**Comment by Szymon-dziewonski** [comment]
I found it problematic, since turbolinks cannot handle `rel="preload"` as @Enalmada mentioned. For me what is annoying, is that after e.g 4 redirect(5 clicks to another pages) get 5 times page loaded

Network tab
![image](https://user-images.githubusercontent.com/2566152/81278582-2cb70300-9056-11ea-9249-2162e2497b80.png)

even script is cached I'll have styles visible in dev tools multiple times because now I have 2 tags 

Styles
![image](https://user-images.githubusercontent.com/2566152/81279492-83710c80-9057-11ea-919f-00a453e0a74c.png)
